File: array/209.长度最小的子数组.js
/*
 * @lc app=leetcode.cn id=209 lang=javascript
 *
 * [209] 长度最小的子数组
 */

// @lc code=start
/**
 * @param {number} target
 * @param {number[]} nums
 * @return {number}
 */
var minSubArrayLen = function(target, nums) {
    let len = nums.length
    let left = 0
    let sum = 0
    let res = Number.MAX_VALUE

    for (let right = 0; right < len; right++) {
        sum += nums[right]

        while (sum >= target) {
            let temp = right - left + 1
            if (temp < res) {
                res = temp
            }
            sum -= nums[left]
            left++
        }
    }

    return res == Number.MAX_VALUE ? 0 : res
};
// @lc code=end

File: array/209.长度最小的子数组.java
/*
 * @lc app=leetcode.cn id=209 lang=java
 *
 * [209] 长度最小的子数组
 */

// @lc code=start
class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        int left = 0;
        int l = nums.length;
        int sum = 0;
        int res = Integer.MAX_VALUE;

        for (int right = 0; right < l; right++) {
            sum += nums[right];
            while (sum >= target) {
                int length = right - left + 1;
                if (length < res) {
                    res = length;
                }
                
                sum -= nums[left];
                left++;
            }
        }

        return res == Integer.MAX_VALUE ? 0 : res;
    }
}
// @lc code=end

